### PR TITLE
Asynchronous threading phase2

### PR DESCRIPTION
**Pull request by john-sharratt**
This patch does a series of refactoring to prevent similar bugs to the one that was found in the asynchronous threading patch.

In particular:
- `inner()` is scoped to the wasi module and no longer usable outside
- `inner()` is marked as `unsafe` and any use of it is audited and wrapped with unsafe
- Cloning `WasiEnv` no longer copies the `inner()`
- Rewinding the stack now passes a return value back to the rewound stack rather than using the memory (which is unsafe)
- The after callback on asynchronous threading has been removed and replaced with rewind return values

### CODE REVIEW

File: lib/wasi/src/runtime/task_manager/mod.rs [hunk 1]
```diff
@@ -166,7 +168,7 @@ impl dyn VirtualTaskManager {
     /// pulled from the worker pool that has a stateful thread local variable
     /// After the poller has successed
```
Comment (syrusakbary) [inline lib/wasi/src/runtime/task_manager/mod.rs:169]: Succeeded?

```suggestion
    /// pulled from the worker pool that has a stateful thread local variable
    /// After the poller has succeeded
```

File: lib/wasi-web/src/pool.rs [hunk 2]
```diff
@@ -79,7 +80,7 @@ struct WasmRunCommand {
     snapshot: Option<InstanceSnapshot>,
     trigger: Option<WasmRunTrigger>,
     update_layout: bool,
-    result: Result<(), Errno>,
+    result: Option<Result<Bytes, ExitCode>>,
```
Comment (theduke) [inline lib/wasi-web/src/pool.rs:83]: Can we have a comment here explaining what this `result` is?

Pretty confusing if you don't know the context. (why `Option`, why `Bytes`, ...)

**Review by theduke — COMMENTED** [review]
Why do we need that thread local?

If we really do need it, we can guard against bugs by storing some thread-relevant value in the thread local and panic on a mismatch.
(We can just store the whole WasiThread, since it's clone, and do Arc::ptr_eq for equality)

(rather than making the functions unsafe)

File: lib/wasi/src/state/env.rs [hunk 3]
```diff
@@ -792,7 +828,7 @@ impl WasiEnv {
         (memory, state)
     }
 
-    pub(crate) fn get_memory_and_wasi_state_and_inodes<'a>(
+    pub(crate) unsafe fn get_memory_and_wasi_state_and_inodes<'a>(
```
[the remark this answers is not in the document]
Comment (john-sharratt) [reply lib/wasi/src/state/env.rs:831]: Fixed - added comment

File: lib/wasi/src/state/env.rs [hunk 4]
```diff
@@ -782,7 +818,7 @@ impl WasiEnv {
         self.state.std_dev_get(fd)
     }
 
-    pub(crate) fn get_memory_and_wasi_state<'a>(
+    pub(crate) unsafe fn get_memory_and_wasi_state<'a>(
```
[the remark this answers is not in the document]
Comment (john-sharratt) [reply lib/wasi/src/state/env.rs:826]: Fixed - added comment

File: lib/wasi/src/lib.rs [hunk 5]
```diff
@@ -289,16 +285,21 @@ impl Clone for WasiVFork {
...
+// Contains the result of a rewind operation
+struct RewindResult {
+    memory_stack: Bytes,
+    rewind_result: Bytes,
+}
+
 // Represents the current thread ID for the executing method
 thread_local!(pub(crate) static CALLER_ID: RefCell<u32> = RefCell::new(0));
-thread_local!(pub(crate) static REWIND: RefCell<Option<bytes::Bytes>> = RefCell::new(None));
+thread_local!(pub(crate) static REWIND: RefCell<Option<RewindResult>> = RefCell::new(None));
```
[the remark this answers is not in the document]
Comment (john-sharratt) [reply lib/wasi/src/lib.rs]: I've now removed these thread locals completely.

File: lib/wasi/src/lib.rs [hunk 6]
```diff
@@ -289,16 +285,21 @@ impl Clone for WasiVFork {
...
             env: Box::new(self.env.as_ref().clone()),
             handle: self.handle.clone(),
         }
     }
 }
 
+// Contains the result of a rewind operation
+struct RewindResult {
+    memory_stack: Bytes,
+    rewind_result: Bytes,
```
[the remark this answers is not in the document]
Comment (john-sharratt) [reply lib/wasi/src/lib.rs]: Comments added

File: lib/wasi-web/src/pool.rs (hunk 2 above)
Comment (john-sharratt) [reply lib/wasi-web/src/pool.rs:83]: Comments added

**Review by theduke — APPROVED** [review]
Much cleaner now without the thread locals!

**Pull request merged**